creme/optim/averager.py

```diff
@@ -0,0 +1,83 @@
+import collections
+
+from . import base
+from . import schedulers
+
+
+__all__ = ['Averager']
+
+
+class Averager(base.Optimizer):
+    """An Stochastic weight averaging wrapper around various sgd optimzers.
+
+    Example:
+
+        ::
+
+            >>> from creme import linear_model
+            >>> from creme import metrics
+            >>> from creme import model_selection
+            >>> from creme.optim.schedulers import Cyclic
+            >>> from creme.optim import Adam, Averager
+            >>> from creme import preprocessing
+            >>> from creme import stream
+            >>> from sklearn import datasets
+
+            >>> X_y = stream.iter_sklearn_dataset(
+            ...     dataset=datasets.load_breast_cancer(),
+            ...     shuffle=True,
+            ...     random_state=42
+            ... )
+
+            >>> adam = Adam(lr=Cyclic(1, 0.01, 100))
+            >>> optimizer = Averager(adam)
+            >>> model = (
+            ...     preprocessing.StandardScaler() |
+            ...     linear_model.LogisticRegression(optimizer)
+            ... )
+            >>> metric = metrics.F1()
+
+            >>> model_selection.progressive_val_score(X_y, model, metric)
+            F1: 0.959441
+
+    References:
+        1. `Averaging Weights Leads to Wider Optima and Better Generalization
+            <https://arxiv.org/abs/1803.05407>`_
+
+    """
+
+    def __init__(self, base_optim):
+        self.base_optim = base_optim
+        self.n_iterations = 0
+        self.c = (
+            base_optim.lr.c
+            if isinstance(base_optim.lr, schedulers.Cyclic) else
+            1
+        )
+        self.w_ = collections.defaultdict(float)
+        self.w_swa = collections.defaultdict(float)
+
+    def update_before_pred(self, w: dict) -> dict:
+        return self.base_optim.update_before_pred(w)
+
+    def _update_after_pred(self, w: dict, g: dict) -> dict:
+
+        # Update the weights
+        self.w_ = self.base_optim._update_after_pred(self.w_.copy(), g)
```

Comment:
Shouldn't you be using `w`?

Reply:
Initially I used `w` but didn't get expected good results. Then on inspecting I found that in `_update_after_pred` of any `creme.optim.Optimizer` the param `w` is the same as whatever we returned in the previous iteration(like if in previous iteration `_update_after_pred` returned `[0.1, 0.2, 0.3]` than in next iteration param `w` will have this value). But in the case of `Averager` we don't want this `w` to train our `base_optim` because this is the average of previous `w's`. So I stored the immediate recent weights trained by the `base_optim` in `w_` and in the next iteration passed these `w_` to the `base_optim` rather than the averaged one. So it's like the `base_optim` has no idea it is going to averaged. And I think this is what we intended with `Averager`, to take an optimizer and average it's weight in regular intervals.